packages/http-proxy-resource/src/BearerTokenCache.ts

```diff
@@ -0,0 +1,160 @@
+import { EventEmitter } from "events";
+
+interface FetchConfig {
+  method: "POST";
+  headers: {
+    "Content-Type": "application/json";
+    Authorization: string;
+  };
+  body: string;
+}
+
+export interface FetchFunction {
+  (uri: string, conf: FetchConfig): Promise<FetchResponse>;
+}
+
+interface FetchResponse {
+  status: number;
+  json(): Promise<any>;
+}
+
+interface BearerTokenCacheConfig {
+  fetchFunc: FetchFunction;
+  timeSource: () => number;
+  tokenRefreshSeconds: number;
+  tokenExpirySeconds: number;
+  authxUrl: string;
+}
+
+class NoTokenError extends Error {}
+
+class BearerTokenCacheEntry {
+  private lastRefreshRequestTime: number;
+  private lastSuccessfulRefreshTime: number | null = null;
+  private forceRefresh: boolean = false;
+
+  private currentToken: Promise<string> | null = null;
+  private nextToken: Promise<string> | null = null;
+
+  constructor(
+    private basicToken: string,
+    private conf: BearerTokenCacheConfig,
+    private errorHandler: (err: Error) => void
+  ) {
+    this.lastRefreshRequestTime = Date.now();
+  }
+
+  get token(): Promise<string> {
+    if (this.currentToken === null) {
+      this.startFetchNextToken();
+      this.currentToken = this.nextToken;
+    } else if (
+      this.forceRefresh ||
+      (this.conf.timeSource() - this.lastRefreshRequestTime) / 1_000 >
+        this.conf.tokenRefreshSeconds
+    ) {
+      this.startFetchNextToken();
+    }
+
+    if (!this.currentToken)
+      throw new Error("currentToken was unexpectedly null");
+    return this.currentToken;
+  }
+
+  get secondsSinceLastRefresh(): number {
+    return (
+      (this.conf.timeSource() -
+        (this.lastSuccessfulRefreshTime ?? this.lastRefreshRequestTime)) /
+      1_000
+    );
+  }
+
+  private startFetchNextToken(): void {
+    this.nextToken = this.fetchNextToken();
+    this.nextToken.then(
+      () => {
+        this.currentToken = this.nextToken;
+        this.lastSuccessfulRefreshTime = this.conf.timeSource();
+      },
+      (err) => {
+        this.errorHandler(err);
+        this.forceRefresh = true;
+      }
+    );
+    this.lastRefreshRequestTime = this.conf.timeSource();
+  }
+
+  private async fetchNextToken(): Promise<string> {
+    const response = await this.conf.fetchFunc(
+      `${this.conf.authxUrl}/graphql`,
+      {
+        headers: {
+          "Content-Type": "application/json",
+          Authorization: this.basicToken,
+        },
+        method: "POST",
+        body: JSON.stringify({
+          query: "query { viewer { token(format:BEARER) } }",
```

Comment:
Requesting `Authorization.token` over the GraphQL API requires the requesting authorization to have the ability to read its own secrets (in addition to reading basic information about itself, as all authorizations have).

This change makes the following required:

https://github.com/the-control-group/authx/blob/9d2aba9879202f8cb7f84f6c514993b84cef252c/packages/authx/src/graphql/GraphQLAuthorization.ts#L156-L165

... while only the following is currently required:

https://github.com/the-control-group/authx/blob/9d2aba9879202f8cb7f84f6c514993b84cef252c/packages/authx/src/graphql/GraphQLAuthorization.ts#L126-L130

This is a breaking change that would require careful evaluation of all currently active grants/authorizations, so it's probably best to simply cache the `.access` associated with each cached authorization.